2-collectors/scala-stream-collector/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -233,41 +280,113 @@ class KinesisSink private (
     ()
   }
 
-  // TODO: limit max retries?
+  /**
+   *  Max number of retries is unlimitted, so when Kinesis stream is under heavy load,
+   *   the events accumulate in collector memory for later retries. The fix for this is to use
+   *   sqs queue as a buffer and sqs2kinesis to move events back from sqs queue to kinesis stream.
+   *  Consider using sqs buffer in heavy load scenarios.
+   *
+   */
   def sendBatch(batch: List[(ByteBuffer, String)], nextBackoff: Long = minBackoff): Unit =
-    if (batch.size > 0) {
+    if (batch.nonEmpty) {
       log.info(s"Writing ${batch.size} Thrift records to Kinesis stream ${streamName}")
-      val putData = for {
-        p <- multiPut(streamName, batch)
-      } yield p
 
-      putData onComplete {
+      multiPut(streamName, batch).onComplete {
         case Success(s) => {
           val results = s.getRecords.asScala.toList
           val failurePairs = batch zip results filter { _._2.getErrorMessage != null }
           log.info(
             s"Successfully wrote ${batch.size - failurePairs.size} out of ${batch.size} records"
           )
-          if (failurePairs.size > 0) {
+          if (failurePairs.nonEmpty) {
             failurePairs.foreach(
               f =>
                 log.error(
                   s"Record failed with error code [${f._2.getErrorCode}] and message [${f._2.getErrorMessage}]"
                 )
             )
-            log.error(s"Retrying all failed records in $nextBackoff milliseconds...")
             val failures = failurePairs.map(_._1)
-            scheduleBatch(failures, nextBackoff)
+            maybeSqs match {
+              case Some(sqs) =>
+                log.info(s"Sending all failed records to SQS buffer queue: ${sqs.sqsBufferName}")
+                putToSqs(sqs, failures)
+              case None =>
+                log.error(s"Retrying all failed records in $nextBackoff milliseconds...")
+                log.warn(
+                  s"${failurePairs.size} failed events scheduled for retry (consider setting a SQS Buffer in config.hocon)"
+                )
+                scheduleBatch(failures, nextBackoff)
+            }
           }
         }
         case Failure(f) => {
           log.error("Writing failed.", f)
-          log.error(s"Retrying in $nextBackoff milliseconds...")
-          scheduleBatch(batch, nextBackoff)
+          maybeSqs match {
+            case Some(sqs) =>
+              log.info(
+                s"Sending all (${batch.size}) events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+              )
+              putToSqs(sqs, batch)
+            case None =>
+              log.error(s"Retrying in $nextBackoff milliseconds...")
+              log.warn(
+                s"All ${batch.size} events from a batch scheduled for retry (consider setting a SQS Buffer in config.hocon)"
+              )
+              scheduleBatch(batch, nextBackoff)
+          }
```

Comment:
GH didn't put it into my review, sorry for double messages.
We expect these two blocks to behave the same, as in log and pass the messages. Extracting a parameterized method might prove useful.

Reply:
thanks for spotting the same logic in almost the same code!